File: fe/fe-core/src/test/java/org/apache/doris/common/CIDRTest.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.common;

import org.junit.Assert;
import org.junit.Test;

public class CIDRTest {
    @Test
    public void testWrongFormat() {
        // no mask
        try {
            CIDR cidr = new CIDR("192.168.17.0/");
            // should not be here
            Assert.assertTrue(false);
        }  catch (Exception e) {
            Assert.assertTrue(e instanceof IllegalArgumentException);
        }

        // mask is too big
        try {
            CIDR cidr = new CIDR("192.168.17.0/88");
            // should not be here
            Assert.assertTrue(false);
        }  catch (Exception e) {
            Assert.assertTrue(e instanceof IllegalArgumentException);
        }

        // ip is too short
        try {
            CIDR cidr = new CIDR("192.168./88");
            // should not be here
            Assert.assertTrue(false);
        }  catch (Exception e) {
            Assert.assertTrue(e instanceof IllegalArgumentException);
        }
    }

    @Test
    public void testNormal() throws UserException {
        // the real value is 10.1.16.0/20
        CIDR cidr = new CIDR("192.168.17.0/20");
        Assert.assertEquals("192.168.17.0", cidr.getIP());
        Assert.assertEquals("255.255.240.0", cidr.getNetmask());
        Assert.assertEquals("192.168.16.0/20", cidr.getCIDR());
        System.out.println("range: " + cidr.getIpRange());
    }

    @Test
    public void testContain() {
        CIDR cidr = new CIDR("192.168.17.0/16");
        Assert.assertTrue(cidr.contains("192.168.88.88"));
        Assert.assertFalse(cidr.contains("192.2.88.88"));
    }
}

File: be/test/util/cidr_test.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "util/cidr.h"

#include <gtest/gtest.h>
#include <stdio.h>
#include <stdlib.h>

#include <iostream>

#include "common/configbase.h"
#include "util/cpu_info.h"
#include "util/logging.h"

namespace doris {

TEST(CIDR, wrong_format) {
    CIDR cidr;
    EXPECT_FALSE(cidr.reset("192.168.17.0/"));
    EXPECT_FALSE(cidr.reset("192.168.17.0.1/20"));
    EXPECT_FALSE(cidr.reset("192.168.17.0.1/88"));
    EXPECT_FALSE(cidr.reset("192.168.17/"));
    EXPECT_FALSE(cidr.reset("192.168.a.a/24"));
    EXPECT_FALSE(cidr.reset("192.168.a.a/16"));
    EXPECT_FALSE(cidr.reset("192.168.0.0/a"));
}

TEST(CIDR, normal) {
    CIDR cidr;
    EXPECT_TRUE(cidr.reset("192.168.17.0/20"));
    EXPECT_TRUE(cidr.reset("10.10.10.10/20"));
    EXPECT_TRUE(cidr.reset("10.10.10.10/32"));
    EXPECT_TRUE(cidr.reset("10.10.10.10"));
}

TEST(CIDR, contains) {
    CIDR cidr;
    EXPECT_TRUE(cidr.reset("192.168.17.0/16"));
    EXPECT_TRUE(cidr.contains("192.168.88.99"));
    EXPECT_FALSE(cidr.contains("192.2.88.99"));
}

} // end namespace doris

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    doris::CpuInfo::init();
    return RUN_ALL_TESTS();
}
